repair.py

```diff
@@ -0,0 +1,162 @@
+            shutil.copyfileobj(fp, temp_file)
+            temp_file.flush()
+            cr_pkginfo = cr.package_from_rpm(temp_file.name)
+
+        new_package = Package.createrepo_to_dict(cr_pkginfo)
+        old_package = ca.content.rpm_package  # not efficient, but eh, don't care.
+
+        old_package.files = new_package["files"]
+        old_package.changelogs = new_package["changelogs"]
+
+        packages_to_update.append(old_package)
+
+    if len(packages_to_update) >= 50:
+        Package.objects.bulk_update(packages_to_update, fields=["files", "changelogs"])
+        total_packages_repaired += len(packages_to_update)
+        packages_to_update.clear()
+        update_total()
+else:
+    Package.objects.bulk_update(packages_to_update, fields=["files", "changelogs"])
+    total_packages_repaired += len(packages_to_update)
+    packages_to_update.clear()
+    update_total()
+
+# Repair on-demand Packages
+
+max_concurrent_downloads = 6
+packages_to_update = []
+
+
+async def repair_on_demand_content(content, remote_artifacts):
+    import aiofiles
+    import aiofiles.os
+    import createrepo_c as cr
+    from pulp_rpm.app.models import Package
+
+    async with aiofiles.tempfile.TemporaryDirectory():
+        for remote_artifact in remote_artifacts:
+            remote = remote_artifact.remote
+            downloader = remote.get_downloader(remote_artifact=remote_artifact)
+            download_result = await downloader.run()
+
+            cr_pkginfo = cr.package_from_rpm(download_result.path)
+            new_package = Package.createrepo_to_dict(cr_pkginfo)
+            old_package = content.rpm_package  # not efficient, but eh, don't care.
+            await aiofiles.os.remove(download_result.path)
+
+            old_package.files = new_package["files"]
+            old_package.changelogs = new_package["changelogs"]
+
+            return old_package
+
+
+for ca in broken_content_artifacts_without_files.select_related("content").iterator():
+    remote_artifacts = list(ca.remoteartifact_set.all().select_related("remote"))
+    package = repair_on_demand_content(ca.content, remote_artifacts)
+    packages_to_update.append(package)
+    if len(packages_to_update) >= 10:
+        packages_to_update = loop.run_until_complete(asyncio.gather(*packages_to_update))
+        Package.objects.bulk_update(packages_to_update, fields=["files", "changelogs"])
+        total_packages_repaired += len(packages_to_update)
+        packages_to_update.clear()
+        update_total()
+else:
+    packages_to_update = loop.run_until_complete(asyncio.gather(*packages_to_update))
+    Package.objects.bulk_update(packages_to_update, fields=["files", "changelogs"])
+    total_packages_repaired += len(packages_to_update)
+    packages_to_update.clear()
+    update_total()
+
+print("...FINISHED")
+
+print(
+"""
+The following repositories need to be re-published:
+"""
+)
+
+repo_versions = RepositoryVersion.objects.with_content(broken_packages)
```

Comment:
This part isn't finished yet.  I'm not sure what information we should give them - which repository versions need to be re-published, or which distributions are affected, how it should be formatted, etc.